rasa/core/training/dsl.py

```diff
@@ -40,13 +44,25 @@ def _parse_item(self, line: Text) -> Optional["Message"]:
             DOCS_BASE_URL
         )
 
-        item_regex = re.compile(r"\s*(.+?):\s*(.*)")
+        # Match three groups:
+        # 1) The correct intent
+        # 2) Optional entities
+        # 3) The message text
+        item_regex = re.compile(r"\s*(.+?)({.*})*:\s*(.*)")
```

Comment:
I noticed this regex matches `hello{: world` (the `hello{` is consumed by the first group, `(.+?)`). Maybe we could modify it a bit so that the `{}` need to be balanced:

`\s*([^{}]+?)({.*})*:\s*(.*)`

I only changed the `.` in the first group for `[^{}]`.

Reply:
I noticed that the regexp does not match the example you provided in the changelog: `greet: /greet{"name": "John"}`. I might be getting two different things confused, though.

Reply:
> I noticed that the regexp does not match the example you provided in the changelog: greet: /greet{"name": "John"}. I might be getting two different things confused, though.

I just tried mine and yours and it works fine. `greet: /greet{"name": "John"}` should be `greet` (group 1), `` (group 2) and then `/greet{"name": "John"}` (group 3)

Reply:
Sorry, I had tested it with `/greet{"name": "John"}` only (thought the `{"name": "John"}` was supposed to be group 2).

Reply:
group 2 would only kick in when you do `greet{"name": "John"}: /greet{"name": "John"}` (for whatever reason 😄 )

Reply:
> I noticed this regex matches hello{: world (the hello{ is consumed by the first group, (.+?)). Maybe we could modify it a bit so that the {} need to be balanced:

> \s*([^{}]+?)({.*})*:\s*(.*)


But in this case it wouldn't recognize anything in `hello{: world` @federicotdn 🤔

Reply:
But do we consider `hello{: world` to be valid? I imagine it would be more probable that the user has made a mistake in that case and forgot to add `}`.

Reply:
so, we raise an Exception in that case?

Reply:
done 👍

Reply:
Thanks - I think an Exception is a good option. As the training data format starts getting more complex I guess it makes sense to inform the user if they have made a mistake.